### ISSUE TITLE
Exception using `.get_ipc_handle(...)` on array from `as_cuda_array(...)

### ISSUE THREAD

**Issue by jakirkham** [numba/numba#5429]
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

<hr>

Am trying to access `.get_ipc_handle(...)` on a Numba `DeviceNDArray` created through `as_cuda_array(...)`. Here's some code using CuPy that demonstrates this use case:

```python
import cupy
import numba.cuda

a = cupy.arange(5)
a2 = numba.cuda.as_cuda_array(a)

a2.get_ipc_handle()
```

However am getting the following exception when calling `.get_ipc_handle(...)`:

```python
Traceback (most recent call last):
  File "script.py", line 7, in <module>
    a2.get_ipc_handle()
  File "/datasets/jkirkham/miniconda/envs/rapids13dev/lib/python3.7/site-packages/numba/cuda/cudadrv/devicearray.py", line 298, in get_ipc_handle
    ipch = devices.get_context().get_ipc_handle(self.gpu_data)
  File "/datasets/jkirkham/miniconda/envs/rapids13dev/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 847, in get_ipc_handle
    memory.owner.handle,
AttributeError: 'cupy.core.core.ndarray' object has no attribute 'handle'
```

It appears this code that is trying to compute the `offset` causes the issue

https://github.com/numba/numba/blob/cc7e7c7cfa6389b54d3b5c2c95751c97eb531a96/numba/cuda/cudadrv/driver.py#L850

cc @Iroy30 @kkraus14

**Comment by stuartarchibald** [comment]
@gmarkall any chance you could take a look at this please? Thanks.

**Comment by gmarkall** [comment]
Already looking into it :-)

**Comment by stuartarchibald** [comment]
> Already looking into it :-)

Brilliant, thanks!

**Comment by gmarkall** [comment]
The use of the owner of a `MemoryPointer` is a bit overloaded:

1. It's used to keep the owning object alive by holding a reference to it - as is set up by `as_cuda_array`.
2. It's also expected to be a `MemoryPointer` which points to the base of the allocation - this made sense when Numba was "isolated" and didn't communicate with other libraries through the CUDA array interface, but doesn't make as much sense now.

I think the second use can be replaced with a call to `cuMemGetAddressRange` instead, to disentangle these two uses, and also allow getting the base of the allocation for pointers that come in through the CUDA array interface - I'm experimenting with a fix now.

**Comment by gmarkall** [comment]
Fix proposed in #5437.

**Comment by Iroy30** [comment]
@jakirkham Thanks for creating the Issue. It is related to the error described in https://github.com/rapidsai/cugraph/issues/790. The proposed fix solves it.
